### ISSUE TITLE
Please drop dependency on the PLUS pack

### ISSUE THREAD

**Issue by GitMensch** [spgennard/vscode_cobol#137]
I know it was added by design from 8dcfae378e4b4fab7318e314cd3af19b57d848b5 and follow-ups. But as the plus pack has its own dependency to more plugins and none of these are necessary for COBOL - please do it the other way around: let the PLUS pack depend on this extension or move the COBOL parts out to a separate extension (I actually would prefer a "Visual Studio Code Extension for COBOL" extension - non mainframers and non-MicroFocus users don't need JCL or MF-directives).

Thank you,
Simon

**Comment by spgennard** [comment]
The COBOL extension provides hlasm support in the source outline view, so it needs something to
handle the opening of these files.

This was part of the driver for including just the syntax support without full hlasm handling and
I thought if you wanted full hlasm support you could include someone else's extension, this
would override the extension handling and add new features provided by there lsp/extension.

If the extension was broken down into multiple extensions, I am unsure it could ever work when 
working with scenarios that require the use of two or more of the sub-divided extension.  

Which would be problematic for someone trying to modernise the code or has two or more types
of code in the same workspace.

**Comment by GitMensch** [comment]
As long as JCL and hlasm are in a separate extension (in this case the plus pack) and this extension adds syntax highlighting and all features then this should be no issue - should it?
I'd really prefer a "pure" COBOL version and having all different stuff done with an extending extension (similar to https://github.com/spgennard/vscode_custom_cobol which works the same way).

The are hundreds of mainframers out there and thousands of non-mainfraimers inspecting COBOL and working *only* on PC - without any special assembler or JCL at all. It would be most useful if both can use the appropriate extension and get what they've asked for, not more (and actually that's the vscode guideline "all extensions should try to do exactly one thing right"; and as people also regularly requests "but I want to install only one thing, the rest should be done under the hood" they've added the extension bundles - maybe that's a useful way to go with COBOL + JCL + HLASM, too?

**Comment by spgennard** [comment]
Using multiple extensions increases the memory footprint and the
development time.   So it's not for me.

In an ideal world having one extension per purpose is nice but that is not
how this has evolved or will continue to be evolved (sorry).

The name of the extension is misleading.

The best I can do is look at changing the name to ensure it is reflective
of the features but this might mean removing this extension as I don't
believe I can change it.

**Comment by GitMensch** [comment]
> Using multiple extensions increases the memory footprint [...]

Obviously another reason for "COBOL-only" people to not have 3 extensions installed.

> The best I can do is look at changing the name to ensure it is reflective
of the features but this might mean removing this extension as I don't
believe I can change it.

I agree.

... actually I suggest a different approach:

* create a branch "cobol",  ideally from [6.2.36](https://github.com/spgennard/vscode_cobol/releases/tag/6.2.36)
* branch "cobol": adjust package.json to remove the recommends part (if you did not create the branch from the tag also undo newer changes)
* branch "cobol": adjust README.md
* adjust master to have an adjusted extension name
* publish both extensions, now with different names and a different feature set

This would provide extensions for both cases and allows to merge the related changes to the COBOL-branch from time to time. @spgennard What do you think?

**Comment by spgennard** [comment]
Following the one-purpose per extension, it would need to be broken down via the different textmate grammars:

For example:
 - GnuCOBOL
 - ACU
 - MF/Generic
 - IBM Enterprise

None of these would be able to co-exist together easily, as only the last one installed would own the extension pool.

Then onto the other grammars:
 - JCL
 - BMS
 - Pl/I
 - dir files

and then the "source view" would need a separate embedded extension, used by each of the above.

Then if I wanted to share source between them I would need either to create libraries for the shared features or an hidden extension service especially if the symbol cache is to be shared between the extensions.

This is a lot of work with little reward, compared to adding extra features/bug fixes.

**Comment by GitMensch** [comment]
I see. But what about the branch which has only the COBOL-Types in and separating the "contained" and "not contained" this way?

... thinking again:
As you want to merge the plus pack in any case I'd suggest a different approach:

* create a fork of this repo (the GitHub support can do this, I can take on the negotiations with them)
* in this fork manually integrate the plus-pack completely (and use its package identifier)
This way the cobol-repo would be "pure" and the plus pack "complete", both with separated code bases but an easy "pull" to get the changes from "pure" into "full"

**Comment by GitMensch** [comment]
The current dependency got more lightweight by not including the broadcommfd.hlasm-language-support dependency (is this temporary/final?); still I'd like to see the plus pack depend on this extension,, not the other way around - is this possible?

ping @spgennard do you see the outlined option (a separate "fork" repo or a separate "COBOL-only" branch) to be possible or have a different option in mind that would allow for a COBOL-only plugin?

**Comment by spgennard** [comment]
The purpose of the other languages is that they are linked to the COBOL.

At some point, the JCL will have support to navigate to symbols references in the project, the COBOL will be able to jump to hlasm symbols and so in.

Separating them out will not allow me to todo this.

COBOL is very rarely used by itself, it always used with other things.

The overhead of supporting the other languages adds <= 64k, so I can't see why it is a problem.

**Comment by GitMensch** [comment]
So hlasm will get back in "some day" and there's no option to change the depdency order between this and the plus pack (which therefore is likely to vanish sooner or later with a full inclusion to come), correct?

Maybe it would be still reasonable to integrate everything into the plus-pack to use its extension name instead of "cobol" only?
Without any kind of sarcasm; What about adjusting the extension name to something like "Everything COBOL-related (COBOL, JCL, HLASM, MF directive and test files, ...)"?

**Comment by spgennard** [comment]
I have already started work on the hlasm and it is currently a grand old size of 6201 bytes.

```shell
-rw-rw-r-- 1 spg spg 6201 May 18 21:46 hlasm.tmLanguage.json
```

EXEC SQL is already supported and is delegated to "source.sql" with some tweaks to handle fields.

EXEC CICS is already particularly supported and my plan is to move this to a separate language, so it can be shared with hlasm, pl/il.

EXEC HTML is also supported too.

The display name for the extension is already:  "COBOL, JCL, PL/I & DIR syntax support"

The name of extension cannot be changed, so, unfortunately, it will also be called COBOL

hlasm, will be added to the display name when I feel it is reasonability complete.

**Comment by GitMensch** [comment]
Can you please add the `EXEC` information to README and possibly display name? I've asked because I did not found the information.

As you seem to implement hlasm on your own instead of pulling another extension in I'm less concerned about it.
Is there a reasonable use for the plus pack as standalone? If not it would be good to "just" integrate it completely, which would also be a good state to close the issue as solved (all external dependencies would be removed).

**Comment by spgennard** [comment]
I cannot think of a reason why you would want to use the plus pack without the main extension,
which is why I am thinking about merging in back in, which would reduce the overall size.

I will look at doing this.

**Comment by GitMensch** [comment]
> I will look at doing this.

Sounds good. I suggest to keep this issue open until then and add a fix note to the commit that actually does this :-)

**Comment by GitMensch** [comment]
Actually fixed with 0c46dbb89dc208497de3feb13c2ad14af410d9f0 and 30a9c0f77ea4b399819e720c1323e53783977dae - thank you.

**Comment by GitMensch** [comment]
Looks like a wrong label, it actually is fixed :-)
